Issue 46

alexferreiradev opened the pull request:
Foi adicionado testes de unit. Rodar todos e verificar se tem falha em algum. Validar se tem alguma melhoria de codigo a fazer.

alexferreiradev committed:
#46 add: teste para leitor de csv

- add teste para leitor de csv com 100% cobertura
- add arquivos para testes
- uso de teste para poder ser independente de OS mesmo com uso de path

alexferreiradev committed:
#46 update: leitor de csv com tratamento de erro

- add tratamento de erro com logger e exception para leitura de csv
- remocao de constantes nao utilizadas

alexferreiradev committed:
#46 update: gitignore sem csv

- alteracao para permitir subir arquivos csv que estejam fora da raiz do projeto

alexferreiradev committed:
#46 update: teste de serviço de certificado

- add verify para logger chamando info
- add mock para repository
- add verify para validar que chamou o persist com a lista de certificados

alexferreiradev committed:
#46 update: command de gerador com javadoc

- add javadoc para nao violar codestyle

joaomarcosribeiroc commented on backend/src/test/java/dev/gojava/module/certificado/service/reader/CsvParticipantReaderTest.java:
```diff
@@ -0,0 +1,102 @@
+        ReaderCommand readerCommand = service.createReaderCommandOrThrow(fakeCsvFile);
+        fakeCsvFile.deleteOnExit();
+
+        MatcherAssert.assertThat(readerCommand.absoluteFilePath, matchesRegex(".*/tempEvento_\\d+.csv"));
```
Esse teste está considerando que o testador está rodando os testes em Linux, uma vez que considera que há a barra / no caminho do arquivo. Os testes deveria ser independentes do OS? Embora na maioria das vezes irá ser testado em linux no CI. Inclusive esse testes não passou no meu sistema win.

alexferreiradev replied:
Na verdade não, pois este caminho é a JVM que cria, e ela padronizou desta forma, mesmo no windows será este path. A mesma coisa acontece pele acesso a arquivos pelo getResources(), começa por / para indicar inicio do resources.

joaomarcosribeiroc replied:
```[INFO] Running dev.gojava.module.certificado.service.reader.CsvParticipantReaderTest
[ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.517 s <<< FAILURE! - in dev.gojava.module.certificado.service.reader.CsvParticipantReaderTest
[ERROR] dev.gojava.module.certificado.service.reader.CsvParticipantReaderTest.givenCsvArquivo_whenCreateReader_thenCriaCommandComArquivoTemp  Time elapsed: 0.008 s  <<< FAILURE!
java.lang.AssertionError:

Expected: a string matching the pattern <.*/tempEvento_\d+.csv>
     but: the string was "C:\\Users\\marco\\AppData\\Local\\Temp\\tempEvento_14883392181639300053.csv"
        at dev.gojava.module.certificado.service.reader.CsvParticipantReaderTest.givenCsvArquivo_whenCreateReader_thenCriaCommandComArquivoTemp(CsvParticipantReaderTest.java:48)

[INFO] 
[INFO] Results:
[INFO]
[ERROR] Failures: 
[ERROR]   CsvParticipantReaderTest.givenCsvArquivo_whenCreateReader_thenCriaCommandComArquivoTemp:48 
Expected: a string matching the pattern <.*/tempEvento_\d+.csv>
     but: the string was "C:\\Users\\marco\\AppData\\Local\\Temp\\tempEvento_14883392181639300053.csv"
[INFO]
[ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  20.823 s
[INFO] Finished at: 2021-09-19T11:33:42-03:00
[INFO] ------------------------------------------------------------------------
```

Esse é o erro que estou tendo.
@alexferreiradev Como resolveríamos isso?

alexferreiradev replied:
Acho que terei que colocar um Path lah para ficar independente de plataforma, pois o absolute leva em consideração o SO... :( dessa eu não esperava, pois na hora de criar o File, ele nao leva em consideração. 

https://docs.oracle.com/javase/7/docs/api/java/io/File.html

alexferreiradev replied:
Vou usar este: https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#isSameFile-java.nio.file.Path-java.nio.file.Path-

alexferreiradev replied:
Alterei para usar o separador do sistema para o teste, não necessitando alterar a implementação, somente o teste.

joaomarcosribeiroc commented on backend/src/test/java/dev/gojava/module/certificado/service/reader/CsvParticipantReaderTest.java:
```diff
@@ -0,0 +1,102 @@
+        File fakeCsvFile = Files.createTempFile("temp_OpenCertificateTest_createCommand", ".csv").toFile();
+        ReaderCommand readerCommand = service.createReaderCommandOrThrow(fakeCsvFile);
+        fakeCsvFile.deleteOnExit();
+
```
Talves seja interessante verificar se o arquivo realmente foi criado com `Assertions.assertTrue(new File(readerCommand.absoluteFilePath).exists());`. Mas estou com uma leve impressão de que isso fugiria do conceito de teste unitário.

alexferreiradev replied:
Seria uma forma a mais. Se eu tivesse colocado, não estaria errado, mas a falta dele, não é um problema, pois continua cobrindo 100% de linhas. Mas irei subir esta correção, pois deixa o teste mais forte, pois pode criar o path para o arquivo temporario e nao criar o arquivo exatamente.

alexferreiradev replied:
Iria até subir, mas o command não tem File, somente o path, então o teste não deve validar a existencia do arquivo, pois a saida é somente o caminho para o arquivo, quem utilizar esta saída que terá realmente a responsabilidade de verificar se o arquivo existe, ai no teste desta classe que teria esta validação. Então você estava certo, foge do escopo deste teste de unit.

joaomarcosribeiroc reviewed (commented):
Iria questionar por que todos os testes da classe CsvParticipantReader não estavam sendo testados, no entato eles são testados inderetamente através dos métodos públicos.

alexferreiradev committed:
#46 update: pom para ignorar teste de integracao

- remocao de testes de integracao do maven test
- estava gerando erro quando ambiente de banco não esta online.
- re-ativar quando alterar para banco em memoria para teste de integracao

alexferreiradev commented:
Não foi corrigido, e para não confudir usuarios do grupo gojava, fechei até ter novos commits.

alexferreiradev commented:
Foi removido contribuidores do projeto, então nao teremos revisores. Irei forçar o merge para o projeto andar.
